### ISSUE TITLE
Allow public/anonymous access to Sandstorm apps

### ISSUE THREAD

**Issue by brylie** [sandstorm-io/sandstorm#1680]
This may be against the grain, but how do I allow general access to a Sandstorm app? Some examples would be:

* A public kanban board so that the general public can observe a software development process
* A public Telescope app, so that the general public can view posted content and possibly sign in to contribute

Ghost is the only Sandstorm app, so far, I have figured out how to make publicly accessible (e.g. via a custom domain with anonymous access).

**Comment by kentonv** [comment]
Hi Brylie,

You can make any app publicly-accessible by:
1. Click "share access" in the top bar.
2. Choose the "get sharable link" tab.
3. Generate a link with the desired permissions.
4. Post it somewhere public (or map your domain to redirect to the link).

Let me know if that covers your use case or if you had something else in mind.

**Comment by brylie** [comment]
One of the goals is to have clean URLs. The generated URL structure is a but complicated.

**Comment by kentonv** [comment]
Ah, OK. We plan at some point to make it possible to assign short URLs to grains as part of sharing.

**Comment by zenhack** [comment]
Wanted to make the point: short != clean. e.g. with github I can always just do:

    git clone https://github.com/<user>/<repo>

So I barely even need to use the website half the time, if I know who I'm collaborating with. "short" doesn't necessarily give me this, e.g. what you get from a URL shortener is still hard to remember.

In this case "Clean" is really what I care about; I want to be able to e.g. host a public git repo and have the UX for people who interact with it not be awkward.

**Comment by brylie** [comment]
Same here. I would like to use Sandstorm to easily maintain public-facing applications.

**Comment by mitar** [comment]
@zenhack: I completely agree. See my (closed) issue on a similar note #1845 and a discussion there.

**Comment by logicfish** [comment]
I'd just like to be able to create static endpoints for my grain that are domains. I'd like these to be routed to the web port of my app, and I would use a virtual host for each endpoint. I'd also like to be able to open ports on the endpoints and have these forwarded to my grains.

**Comment by zenhack** [comment]
Quoting Mark Fisher (2017-11-26 05:59:55)

> I'd just like to be able to create static endpoints for my grain that
> are domains. I'd like these to be routed to the web port of my app,
> and I would use a virtual host for each endpoint.

The way I envisioned this working is that the workflow would be similar
to creating a share link, but instead of a URL with an unguessable token
in it, it would just be a human-readable subdomain. E.g. I could make a
gogs grain with my repositories, and then generate a "public" share link
as `gogs.pub.me.sandcats.io`.

> I'd also like to be able to open ports on the endpoints and have these
> forwarded to my grains.

By ports you mean tcp/udp ports? I suppose you could do this for TLS
connections with SNI, but it's not possible to dispatch on hostname at
the TCP level, since it doesn't exist there.

It is currently possible for apps to request the ability to listen on a
port, but (a) you need to be an administrator to grant the request, (b)
it doesn't hook into the domain name/grain dispatching at all -- you
just have the port on the sandstorm box. It also requires some
sandstorm-specific log inside the app. The logic is described here:

    https://docs.sandstorm.io/en/latest/developing/powerbox/#special-powerbox-request-types

The interface you'd want in order to listen is IpInterface. I don't know
of any apps that have used this though.

**Comment by Osndok** [comment]
I feel like this is really needed.

It might appear a little janky, but in terms of cost/benefit... I recommend that we simply provide a **default presentation text** (visible to all users & visitors; settable in the admin panel)... possibly with it's own side-bar tab so you can get back to it ("Home"?)... where the administrator can put a welcome message and a list of grain share-links that they would like to have publicly accessible.

**Comment by logicfish** [comment]
I did mean tcp/udp endpoints. would it not be possible to replace the static server with a proxy router to the grain? maybe have a route table where the grain can export the paths it wants to map.

**Comment by zenhack** [comment]
Quoting Mark Fisher (2019-03-21 23:39:45)

>    I did mean tcp/udp endpoints. would it not be possible to replace the
>    static server with a proxy router to the grain? maybe have a route
>    table where the grain can export the paths it wants to map.

I'm not entirely sure I follow. Paths are an HTTP notion; they don't
exist at the level of TCP/UDP. If you want to do something smarter than
just handing the grain a whole TCP port, you need some way for the
server to look at the connection and decide how to route it. None of the
information sandstorm currently uses exists at that layer, and I'm not
sure what you'd use instead.

If you wanted to restrict to TLS connections, you could perhaps use SNI
to make the routing decision, but I don't know how you'd do just raw TCP
or UDP.

**Comment by logicfish** [comment]
I was referring to two separate situations, one about the public web access points, which it would be nice to have a reverse proxy to the grain setup from the main server configuration, and the other point was allowing a forwarded port to the grain.

**Comment by zenhack** [comment]
Quoting Mark Fisher (2019-04-05 06:34:46)

>    I was referring to two separate situations, one about the public web
>    access points, which it would be nice to have a reverse proxy to the
>    grain setup from the main server configuration, and the other point was
>    allowing a forwarded port to the grain.

Aha, that makes much more sense.
